In [1]:
import requests

In [2]:
class Rate:

    def __init__(self, format_='value'):
        self.format = format_
    
    def exchange_rates(self):
        """
        Возвращает ответ сервиса с информацией о валютах в виде:
        
        {
            'AMD': {
                'CharCode': 'AMD',
                'ID': 'R01060',
                'Name': 'Армянских драмов',
                'Nominal': 100,
                'NumCode': '051',
                'Previous': 14.103,
                'Value': 14.0879
                },
            ...
        }
        """
        self.r = requests.get('https://www.cbr-xml-daily.ru/daily_json.js')
        return self.r.json()['Valute']
    
    def make_format(self, currency):
        """
        Возвращает информацию о валюте currency в двух вариантах:
        - полная информация о валюте при self.format = 'full':
        Rate('full').make_format('EUR')
        {
            'CharCode': 'EUR',
            'ID': 'R01239',
            'Name': 'Евро',
            'Nominal': 1,
            'NumCode': '978',
            'Previous': 79.6765,
            'Value': 79.4966
        }
        
        Rate('value').make_format('EUR')
        79.4966
        """
        response = self.exchange_rates()
        
        if currency in response:
            if self.format == 'full':
                return response[currency]
            
            if self.format == 'value':
                return response[currency]['Value']
        
        return 'Error'
    
    def eur(self):
        """Возвращает курс евро на сегодня в формате self.format"""
        return self.make_format('EUR')
    
    def usd(self):
        """Возвращает курс доллара на сегодня в формате self.format"""
        return self.make_format('USD')
    
    def brl(self):
        """Возвращает курс бразильского реала на сегодня в формате self.format"""
        return self.make_format('BRL')

In [3]:
r = Rate(format_='full')

In [4]:
r.eur()

{'ID': 'R01239',
 'NumCode': '978',
 'CharCode': 'EUR',
 'Nominal': 1,
 'Name': 'Евро',
 'Value': 78.4431,
 'Previous': 77.7965}

In [5]:
r = Rate(format_='value')

In [6]:
r.eur()

78.4431

# Задание 1
Напишите функцию, которая возвращает название валюты (поле ‘Name’) с максимальным значением курса с помощью сервиса https://www.cbr-xml-daily.ru/daily_json.js

In [7]:
import requests

In [8]:
class Rate_2:

    def __init__(self, format_='value'):
        self.format = format_
    
    def exchange_rates(self):
        self.r = requests.get('https://www.cbr-xml-daily.ru/daily_json.js')
        return self.r.json()['Valute']

In [12]:
class Rate_2(Rate):
    def __init__(self):
        super().__init__(format_='full')
        
    def max_rate(self):
        val_ue = []
        rate = []
#         self.rates = []
        x = self.exchange_rates()
        for key, value in x.items():
            val_ue.append(value['Value'])
#             self.rates.append([key, value['Value']])
            rate.append(key)
#         print(self.val_ue)
#         print(self.rate)
#         print(self.rates)
        return rate[val_ue.index(max(val_ue))] 

In [13]:
f = Rate_2()

In [14]:
f.max_rate()

'XDR'

In [40]:
""" Для красоты выводим курс XDR
"""
class Rate_2:
  
    def __init__(self, format_='value'):
        self.format = format_
    
    def exchange_rates(self):

        self.r = requests.get('https://www.cbr-xml-daily.ru/daily_json.js')
        return self.r.json()['Valute']
    
    def make_format(self, currency):

        response = self.exchange_rates()
        
        if currency in response:
            if self.format == 'full':
                return response[currency]
            
            if self.format == 'value':
                return response[currency]['Value']
        
        return 'Error'
    
    def XDR(self):
         return self.make_format('XDR')
    
class CurrencyCodes(Rate):
    def __init__(self):
        super().__init__(format_='full')

In [41]:
f = Rate_2(format_='value')

In [43]:
f.XDR()

98.1807

# ВОПРОС: 
а можно ли вывести f.max_rate() и сразу же цифру, чтобы не делать допольнительных вышеуказанных формул??
чтобы был такой ответ: 
"XDR" = 98.1807

А то что-то я запутался ))

# Задание 2
Добавьте в класс Rate параметр diff (со значениями True или False), который в случае значения True в методах eur и usd будет возвращать не курс валюты, а изменение по сравнению в прошлым значением. Считайте, self.diff будет принимать значение True только при возврате значения курса. При отображении всей информации о валюте он не используется.

In [111]:
class Rate_3(Rate):
    def __init__(self):
        super().__init__(format_='value')
        
    def eur(self, diff):
        if diff == True:
            return round(self.exchange_rates()['EUR']['Value'] - self.exchange_rates()['EUR']['Previous'], 4)
        else:
            return self.make_format('EUR')
    
    def usd(self, diff):
        if diff == True:
            return round(self.exchange_rates()['USD']['Value'] - self.exchange_rates()['USD']['Previous'], 4)
        else:
            return self.make_format('USD')

In [112]:
f = Rate_3()

In [113]:
f.eur(diff = False)

78.4431

In [114]:
f.usd(diff = False)

71.8804

In [115]:
# разница с предыдущим днём (EUR):
f.eur(diff = True)

0.6466

In [116]:
# разница с предыдущим днём (USD):
f.usd(diff = True)

0.9564

# ВОПРОС:
А как посчитать разницу с этим же днем, но предыдущего месяца? 

# Задание 3
Напишите класс Designer, который учитывает количество международных премий для дизайнеров (из презентации: "Повышение на 1 грейд за каждые 7 баллов. Получение международной премии – это +2 балла"). Считайте, что при выходе на работу сотрудник уже имеет две премии и их количество не меняется со стажем (конечно если хотите это можно вручную менять). Выполните проверку для 20 аккредитаций дизайнера Елены.

In [91]:
class Employee:
    def __init__(self, name, seniority):
        self.name = name
        self.seniority = seniority
        
        self.grade = 1
    
    def grade_up(self):
        """Повышает уровень сотрудника"""
        self.grade += 1
    
    def publish_grade(self):
        """Публикация результатов аккредитации сотрудников"""
        print(self.name, self.grade)
    
    def check_if_it_is_time_for_upgrade(self):
        pass

In [92]:
class Developer(Employee):
    def __init__(self, name, seniority):
        super().__init__(name, seniority)
    
    def check_if_it_is_time_for_upgrade(self):
        # для каждой аккредитации увеличиваем счетчик на 1
        # пока считаем, что все разработчики проходят аккредитацию
        self.seniority += 1
        
        # условие повышения сотрудника из презентации
        if self.seniority % 5 == 0:
            self.grade_up()
        
        # публикация результатов
        return self.publish_grade()

In [93]:
# проверяем как работает система повышения сотрудников на примере отдела разработки

# разработчик Александр только что пришел в компанию
alex = Developer('Александр', 0)

In [94]:
alex.grade, alex.name

(1, 'Александр')

In [119]:
class Designer(Employee):
    def __init__(self, name, seniority, awards):
        super().__init__(name, seniority)
        self.awards = awards
        self.seniority += self.awards * 2
    
    def check_if_it_is_time_for_upgrade(self):
        self.seniority += 1
#         self.seniority += 3 #Для наглядности
        
        if self.seniority % 7 == 0:
            self.grade_up()
            
        return self.publish_grade()

In [120]:
Elena = Designer('Елена', seniority=0, awards=2)

In [121]:
for i in range(20):
    Elena.check_if_it_is_time_for_upgrade()

Елена 1
Елена 1
Елена 2
Елена 2
Елена 2
Елена 2
Елена 2
Елена 2
Елена 2
Елена 3
Елена 3
Елена 3
Елена 3
Елена 3
Елена 3
Елена 3
Елена 4
Елена 4
Елена 4
Елена 4
